### PR TITLE
RUBY-2007 Add AWS KMS support for explicit and auto encryption

### PR DESCRIPTION
**Pull request by egiurleo**

### CODE REVIEW

File: lib/mongo/crypt/encryption_io.rb [hunk 1]
```diff
@@ -86,6 +86,40 @@ def mark_command(cmd)
...
+        ssl_socket = Socket::SSL.new(host, port, host, socket_timeout, Socket::PF_INET)
+        ssl_socket.write(message)
+
+        num_bytes_needed = kms_context.bytes_needed
+
+        while num_bytes_needed > 0
+          bytes = []
+          while !ssl_socket.eof?
+            bytes << ssl_socket.readbyte
+          end
```
Comment (egiurleo) [inline lib/mongo/crypt/encryption_io.rb]: I'm sure there is a more efficient way to do this and would love some guidance.

File: lib/mongo/crypt/encryption_io.rb [hunk 2]
```diff
@@ -86,6 +86,40 @@ def mark_command(cmd)
...
       end
+
+      # Get information about the AWS encryption key and feed it to the the
+      # KmsContext object
+      #
+      # @param [ Mongo::Crypt::KmsContext ] kms_context A KmsContext object
+      #   corresponding to one AWS KMS data key. Contains information about
+      #   the endpoint at which to establish a TLS connection and the message
+      #   to send on that connection.
+      def feed_kms(kms_context)
```
Comment (egiurleo) [inline lib/mongo/crypt/encryption_io.rb]: This is my interpretation of the [integration guide](https://github.com/mongodb/libmongocrypt/blob/master/integrating.md#state-mongocrypt_ctx_need_kms). Keep in mind that I know very little about the `Socket::SSL` module in the driver, so feedback would be appreciated 🙏

File: lib/mongo/crypt/data_key_context.rb [hunk 3]
```diff
@@ -104,16 +118,16 @@ def set_aws_master_key(master_key_opts)
...
+          key,
         )
       end
 
       def set_aws_endpoint(endpoint)
         unless endpoint.is_a?(String)
           raise ArgumentError.new(
             "#{endpoint} is an invalid AWS master_key endpoint. " +
-            "The master_key endpoint option must be a String"
+            "The :endpoint key of the :master_key options Hash must be a String"
```
[the remark this answers is not in the document]
Comment (egiurleo) [reply lib/mongo/crypt/data_key_context.rb]: Thanks!

File: lib/mongo/crypt/encryption_io.rb [hunk 4]
```diff
@@ -86,6 +86,38 @@ def mark_command(cmd)
...
+        endpoint = kms_context.endpoint
+        message = kms_context.message
+
+        socket_timeout = 10
```
[the remark this answers is not in the document]
Comment (egiurleo) [reply lib/mongo/crypt/encryption_io.rb]: No, a value hasn't been set in the spec yet: [https://jira.mongodb.org/browse/SPEC-1394](https://jira.mongodb.org/browse/SPEC-1394)

Go, Java, and Python all use 10s. Do you have a strong preference on what this value should be?

[the remark this answers is not in the document]
Comment (p-mongo) [reply lib/mongo/crypt/encryption_io.rb]: No but could you add a comment referencing the spec ticket please?

**Pull request merged**